### DISCUSSION TITLE
pydeck binary transfer data prep

### DISCUSSION THREAD

**Discussion by cornhundred** [visgl/deck.gl#5828] [Q&A] [answered]
Hi, is there a faster way to prepare data for binary data transfer than using apply (see below code from [docs](https://deckgl.readthedocs.io/en/latest/binary_transfer.html?highlight=binary%20transfer#usage))?

```
nodes["position"] = nodes.apply(lambda row: [row["x"], row["y"], row["z"]], axis=1)
```

It can take ~5min on a dataset with ~millions of rows.

**Comment by ajduberstein** [comment, marked as answer]
Can you try:

`nodes['position'] = nodes[['x', 'y', 'z']].values.tolist()`

`.apply` will be slower. It shouldn't be the recommended path in the docs.

You can get a sense of what pydeck is doing by looking at how deck.gl handles [binary attributes](https://deck.gl/docs/developer-guide/performance#supply-attributes-directly).

**Reply by cornhundred** [reply]
Thanks that is much faster
